### DISCUSSION TITLE
UserInfo endpoint not fully standards compliant

### DISCUSSION THREAD

**Discussion by dteleguin** [keycloak/keycloak#8801] [Ideas]
The UserInfo endpoint currently handles error cases in a way that seems to be not fully standards compliant.

According to OpenID Connect Core 1.0, chapter [5.3 UserInfo Endpoint](https://openid.net/specs/openid-connect-core-1_0.html#UserInfo): 

> The UserInfo Endpoint is an OAuth 2.0 Protected Resource that returns Claims about the authenticated End-User. To obtain the requested Claims about the End-User, the Client makes a request to the UserInfo Endpoint using an Access Token obtained through OpenID Connect Authentication.

> The UserInfo Endpoint MUST accept Access Tokens as OAuth 2.0 Bearer Token Usage [[RFC6750](https://datatracker.ietf.org/doc/html/rfc6750)].

Currently, when handling error cases, the UserInfo endpoint behaves in a manner more similar the Token endpoint rather than an OAuth 2.0 protected resource.

1. When the request lacks authentication information, UserInfo responds with a `400 Bad Request`, no `WWW-Authenticate` header, and a JSON body:

```
HTTP/1.1 400 Bad Request
{
	"error": "invalid_request",
	"error_description": "Token not provided"
}
```

But, according to RFC6750, it SHOULD respond with a `401 Unauthorized`; MUST include the HTTP `WWW-Authenticate` response header field; SHOULD NOT include an error code or other error information:

```
HTTP/1.1 401 Unauthorized
WWW-Authenticate: Bearer realm="example"
```

Among the prominent IdMs that do follow this convention are Ping ([link](https://docs.pingidentity.com/bundle/pingfederate-93/page/bkn1564003025596.html)), Okta ([link](https://developer.okta.com/docs/reference/api/oidc/#response-example-error-6)), Connect2ID ([link](https://connect2id.com/products/server/docs/api/userinfo)), Auth0 and IdentityServer4.

2. In case of any error condition, UserInfo responds with a JSON body containing `error` and `error_description` attributes. This mimics the behavior of the Token endpoint; there is no such requirement for OAuth 2.0 protected resources using bearer token authentication, so this behavior is non-standard. By providing a JSON body for error responses, we encourage the users to depend on a non-standard feature.

3. In case of invalid MTLS binding and/or missing client certificate, the `unauthorized_client` error code is used. This is not recognized as a valid error code for OAuth 2.0 protected resources; most likely this has been borrowed from Token endpoint where it would be indeed a valid error code. Justin Richer of OAuth WG [recommends](https://mailarchive.ietf.org/arch/msg/oauth/XfeH2q0Rwa2YocsR484xk-8LMqc/) to use `invalid_token` instead, both for invalid binding and missing/invalid client certificate. @tnorimat 

Also, a couple of minor issues not directly related to the standards:
- No `openid` scope check: as UserInfo is a feature exclusive to OIDC, it might make sense to reject access tokens that do not have the `openid` scope. IdentityServer4 is one of the IdMs that [is known](https://github.com/IdentityServer/IdentityServer4/issues/1194) to do that.
- No token revocation check: a UserInfo call would succeed with an access token that has been previously revoked.

**Comment by dteleguin** [comment]
Background: the above was discovered while working on DPoP. This feature relies on the ability to correctly inject `WWW-Authenticate` fields and thus is sort of incompatible with what we have now in UserInfo. To me, this looks like a blocker for DPoP. I think I could submit an initial DPoP PR omitting the UserInfo support, so the review process could start while this one is being sorted. WDYT @mposolda ?

**Comment by mposolda** [comment]
@dteleguin @tnorimat +1 to all the points. Thanks for pointing this!

My vote is also to add some note to the "Migration guide" to the documentation in relation to this because some clients can rely on the previous (no specs compliant) behaviour. I don't think that we need to cover backwards compatibility in the codebase as the behaviour is not specs compliant as you pointed. So just documentation is sufficient IMO. I would not add documentation about the last point (Token revocation check) as this one is a bit related to security, but other points (1,2,3 and openid scope) will be nice to document.

**Reply by tnorimat** [reply]
@mposolda I agree with your suggestion for adding some documentation  for backward compatibility

**Comment by dteleguin** [comment]
Created #14184
